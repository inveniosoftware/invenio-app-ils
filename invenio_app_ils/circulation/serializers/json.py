# -*- coding: utf-8 -*-
#
# Copyright (C) 2018-2020 CERN.
#
# invenio-app-ils is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Loan JSON serializers."""

from invenio_records_rest.serializers.json import JSONSerializer

from invenio_app_ils.literature.serializers.custom_fields import \
    field_cover_metadata

from .custom_fields import (field_is_overdue, field_pickup_location,
                            field_transaction_location, field_transaction_user)


class LoanJSONSerializer(JSONSerializer):
    """Serialize loan."""

    def transform_record(self, pid, record, links_factory=None, **kwargs):
        """Transform record into an intermediate representation."""
        loan = super().transform_record(
            pid, record, links_factory=links_factory, **kwargs
        )
        field_is_overdue(loan["metadata"])
        field_pickup_location(loan["metadata"])
        field_transaction_location(loan["metadata"])
        field_transaction_user(loan["metadata"])
        field_cover_metadata(loan["metadata"].get("document", {}))
        return loan

    def transform_search_hit(
        self, pid, record_hit, links_factory=None, **kwargs
    ):
        """Transform search result hit into an intermediate representation."""
        hit = super().transform_search_hit(
            pid, record_hit, links_factory=links_factory, **kwargs
        )
        field_is_overdue(hit["metadata"])
        field_cover_metadata(hit["metadata"].get("document", {}))
        return hit
