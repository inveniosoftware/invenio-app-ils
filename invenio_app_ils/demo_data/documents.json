[
  {
    "title": "An eventful journey to unification of all the fundamental forces",
    "authors": [
      {
        "full_name": "Delbourgo, Robert",
        "roles": ["DATA_CURATOR"]
      }
    ],
    "edition": "2",
    "imprint": {
      "date": "2020",
      "place": "Hackensack, NJ",
      "publisher": "World Scientific"
    },
    "abstract": "Book 2 of the Physics is arguably the best introduction to Aristotle's ideas, as well as being the most interesting and representative book in the whole of his corpus. It defines nature and distinguishes natural science from mathematics. It introduces the seminal idea of four causes, or four modes of explanation. It defines chance, but rejects a theory of chance and natural selection in favour of purpose in nature. Simplicius, writing in the sixth century AD, adds his own considerable contribution to this work. Seeing Aristotle's God as a Creator, he discusses how nature relates to soul, adds Stoic and Neoplatonist causes to Aristotle's list of four, and questions the likeness of cause to effect. He discusses missing a great evil or a great good by a hairsbreadth and considers whether animals act from reason or natural instinct. He also preserves a Posidonian discussion of mathematical astronomy.",
    "languages": ["ENG"],
    "created_by": {
      "type": "script",
      "value": "demo"
    },
    "restricted": false,
    "document_type": "BOOK",
    "cover_metadata": {
      "isbn": "9789811210143",
      "urls": {}
    },
    "identifiers": [
      {
        "scheme": "ISBN",
        "value": "9789811210143"
      }
    ],
    "number_of_pages": "159",
    "publication_year": "2020"
  },
  {
    "title": "Introduction to cryptography",
    "edition": "2x",
    "authors": [
      {
        "full_name": "Knebl, Helmut",
        "roles": ["RESEARCHER"]
      }
    ],
    "imprint": {
      "date": "2002",
      "place": "Berlin",
      "publisher": "Springer"
    },
    "languages": ["ENG"],
    "created_by": {
      "type": "script",
      "value": "demo"
    },
    "restricted": false,
    "document_type": "BOOK",
    "cover_metadata": {
      "isbn": "",
      "urls": {}
    },
    "number_of_pages": "310",
    "publication_year": "2002"
  },
  {
    "title": "Sidelights on relativity",
    "authors": [
      {
        "full_name": "Einstein, Albert",
        "roles": ["RESEARCHER"]
      }
    ],
    "imprint": {
      "date": "1983",
      "place": "New York, NY",
      "publisher": "Dover Publications"
    },
    "abstract": "This Dover edition, first published in 1983, is an unabridged and unaltered republication of the translation by G. B. Jeffrey, D.Sc., and W. Perret, Ph.D., first published by E. P. Dutton and Company, Publishers, N.Y., in 1922.",
    "languages": ["ENG"],
    "edition": "x",
    "created_by": {
      "type": "script",
      "value": "demo"
    },
    "restricted": false,
    "document_type": "BOOK",
    "cover_metadata": {
      "isbn": "9780486245119",
      "urls": {}
    },
    "identifiers": [
      {
        "scheme": "ISBN",
        "value": "9780486245119"
      }
    ],
    "number_of_pages": "56",
    "publication_year": "1983"
  },
  {
    "title": "Big Bang",
    "edition": "3x",
    "authors": [
      {
        "full_name": "Jacob, Irene",
        "roles": ["DATA_CURATOR"]
      }
    ],
    "imprint": {
      "date": "2019",
      "place": "Paris",
      "publisher": "Albin Michel"
    },
    "abstract": "Ici, devant le deuil et la promesse de la vie, je me tiens et j'avance... Irène Jacob, l'inoubliable interprète de La Double vie de Véronique, rapproche deux évènements simultanés et essentiels : la disparition de son père, grand physicien, et la naissance d'un enfant. Ce texte subtil, sensible et tendre lui permet de revenir à ces deux moments où le miracle de la vie et l'infiniment grand se mêlent. La physique, le firmament à perte de vue, le big bang, les questions de la psychanalyse mais aussi la perception du temps, la brume des souvenirs ou les images à vif, les émotions, leurs manifestations explosives, le sommeil et les rêves irisent ce récit tout à la fois bouleversant et lumineux.",
    "languages": ["ENG"],
    "created_by": {
      "type": "script",
      "value": "demo"
    },
    "restricted": false,
    "document_type": "BOOK",
    "cover_metadata": {
      "isbn": "9782226442703",
      "urls": {}
    },
    "identifiers": [
      {
        "scheme": "ISBN",
        "value": "9782226442703"
      }
    ],
    "number_of_pages": "232",
    "publication_year": "2019"
  },
  {
    "title": "Using REDUCE in high energy physics",
    "edition": "4x",
    "authors": [
      {
        "full_name": "Grozin, A G",
        "roles": ["DATA_CURATOR"]
      }
    ],
    "imprint": {
      "date": "2005",
      "place": "Cambridge",
      "publisher": "Cambridge University Press"
    },
    "abstract": "This is an abstract",
    "languages": ["ENG"],
    "created_by": {
      "type": "script",
      "value": "demo"
    },
    "restricted": false,
    "document_type": "BOOK",
    "cover_metadata": {
      "isbn": "9780521019521",
      "urls": {}
    },
    "identifiers": [
      {
        "scheme": "ISBN",
        "value": "9780521019521"
      }
    ],
    "number_of_pages": "384",
    "publication_year": "2005"
  },
  {
    "title": "Proceedings of the 23rd International Workshop on ECR Ion Sources",
    "edition": "5x",
    "authors": [
      {
        "full_name": "Schaa, Volker R W",
        "roles": ["EDITOR"]
      }
    ],
    "imprint": {
      "date": "2019",
      "place": "Geneva",
      "publisher": "JACOW"
    },
    "languages": ["ENG"],
    "created_by": {
      "type": "script",
      "value": "demo"
    },
    "restricted": false,
    "document_type": "PROCEEDINGS",
    "cover_metadata": {
      "isbn": "9783954501960",
      "urls": {}
    },
    "identifiers": [
      {
        "scheme": "ISBN",
        "value": "9783954501960"
      }
    ],
    "conference_info": [
      {
        "year": 2018,
        "dates": "10 - 14 Sep. 2018",
        "place": "Catania",
        "title": "23rd International Workshop on ECR Ion Sources",
        "series": "International Workshop on ECR Ion Sources",
        "country": "IT"
      }
    ],
    "number_of_pages": "43",
    "publication_year": "2019"
  }
]
