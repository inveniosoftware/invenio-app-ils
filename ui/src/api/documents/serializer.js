import isEmpty from 'lodash/isEmpty';
import { fromISO } from '../date';

function serializeResponse(hit) {
  let result = {};
  if (!isEmpty(hit)) {
    result['id'] = hit.id;
    result['created'] = fromISO(hit.created);
    result['updated'] = fromISO(hit.updated);
    if (hit.links) {
      result['links'] = hit.links;
    }
    if (!isEmpty(hit.metadata)) {
      result['metadata'] = hit.metadata;
      if (!isEmpty(hit.metadata.circulation.next_available_date)) {
        result['metadata']['circulation']['next_available_date'] = new Date(
          result['metadata']['circulation']['next_available_date']
        ).toDateString();
      }
      result['pid'] = hit.metadata.pid;
    }
    if (!isEmpty(hit.metadata.imprints)) {
      result.metadata.imprints = hit.metadata.imprints.map(imprint => {
        imprint.date = imprint.date ? fromISO(imprint.date) : null;
        return imprint;
      });
    }
  }
  return result;
}

export const serializer = {
  fromJSON: serializeResponse,
};
