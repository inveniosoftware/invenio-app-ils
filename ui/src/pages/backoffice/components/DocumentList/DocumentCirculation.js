import React, {Component} from 'react';
import {Icon, List} from "semantic-ui-react";
import PropTypes from 'prop-types';

export default class DocumentCirculation extends Component {
  renderOverbookIcon = isOverbooked => {
    const name = isOverbooked ? 'check' : 'minus';
    const color = isOverbooked ? 'red' : 'green';
    return (
      <Icon name={name} color={color} size="small"/>
    );
  };

  render() {
    const {document} = this.props;
    const circulation = document.metadata.circulation;
    return (
      <List verticalAlign="middle" className={'document-circulation'}>
        <List.Item>
          <List.Content floated="right">
            <strong>{document.metadata.items.total}</strong>
          </List.Content>
          <List.Content>total items</List.Content>
        </List.Item>
        <List.Item>
          <List.Content floated="right">
            <strong>{circulation.has_items_for_loan}</strong>
          </List.Content>
          <List.Content>available items</List.Content>
        </List.Item>
        <List.Item>
          <List.Content floated="right">
            <strong>{circulation.active_loans}</strong>
          </List.Content>
          <List.Content>active loans</List.Content>
        </List.Item>
        <List.Item>
          <List.Content floated="right">
            <strong>{circulation.pending_loans}</strong>
          </List.Content>
          <List.Content>pending loans</List.Content>
        </List.Item>
        <List.Item>
          <List.Content floated="right">
            {this.renderOverbookIcon(circulation.overbooked)}
          </List.Content>
          <List.Content>overbooked</List.Content>
        </List.Item>
      </List>
    );
  }
}

DocumentCirculation.propTypes = {
  document: PropTypes.object.isRequired
};
