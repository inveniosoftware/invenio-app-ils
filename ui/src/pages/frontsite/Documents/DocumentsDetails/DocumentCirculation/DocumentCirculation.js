import React, { Component } from 'react';
import { Divider, Header, List, Popup, Segment } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import { LoginRedirectButton } from '@authentication/components';
import { LoanRequestForm } from '../LoanRequestForm';
import { AuthenticationGuard } from '@authentication/components/AuthenticationGuard';
import { ILSImagePlaceholder } from '@components/ILSPlaceholder';

class DocumentEItems extends Component {
  render() {
    const eitems = this.props.document.metadata.eitems.hits.map(eitem => {
      return (
        <List.Item key={eitem.pid}>
          <List.Icon name={'linkify'} />
          <List.Content>
            Read <a href="#TODO">online</a>
          </List.Content>
        </List.Item>
      );
    });
    if (eitems.length > 0) {
      return (
        <>
          <Header as="h3">Access online</Header>
          <List>{eitems}</List>
          <Divider horizontal>Or</Divider>
        </>
      );
    }
    return null;
  }
}

class BookAvailability extends Component {
  render() {
    const circulationData = this.props.document.metadata.circulation;
    if (circulationData.has_items_for_loan > 0) {
      return (
        <List.Item>
          <Popup
            content="Calculated based on current library stock"
            trigger={<List.Icon name={'info'} />}
          />
          <List.Content
            className={
              circulationData.has_items_for_loan > 0
                ? 'text-success'
                : 'text-danger'
            }
          >
            Available for loan <span className={'success'}>now</span>
          </List.Content>
        </List.Item>
      );
    } else if (circulationData.next_available_date) {
      return (
        <List.Item>
          <List.Icon name={'info'} />
          <List.Content>
            Available for loan from <b>{circulationData.next_available_date}</b>
          </List.Content>
        </List.Item>
      );
    } else {
      return (
        <List.Item>
          <List.Icon name={'info'} />
          <List.Content>
            Long waiting queue. Contact us for more information.
          </List.Content>
        </List.Item>
      );
    }
  }
}

export default class DocumentCirculation extends Component {
  loginToLoan = () => {
    return <LoginRedirectButton content={'Login to loan'} />;
  };

  renderLoanRequestForm = () => {
    return <LoanRequestForm document={this.props.documentDetails} />;
  };

  renderEItems = () => {};

  render() {
    const { isLoading, documentDetails } = this.props;
    return (
      <Segment className={'highlighted'}>
        <ILSImagePlaceholder style={{ height: 400 }} isLoading={isLoading}>
          <DocumentEItems document={documentDetails} />
          <Header as="h3" content="Request loan" />
          <List>
            <BookAvailability document={documentDetails} />
          </List>
          <AuthenticationGuard
            authorizedComponent={this.renderLoanRequestForm}
            loginComponent={this.loginToLoan}
          />
        </ILSImagePlaceholder>
      </Segment>
    );
  }
}

DocumentCirculation.propTypes = {
  documentDetails: PropTypes.object.isRequired,
  isLoading: PropTypes.bool.isRequired,
};
