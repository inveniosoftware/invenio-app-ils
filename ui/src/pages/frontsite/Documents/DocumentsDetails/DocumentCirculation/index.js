import { connect } from 'react-redux';
import DocumentCirculationComponent from './DocumentCirculation';

const mapStateToProps = state => ({
  documentDetails: state.documentDetailsFront.data,
  isLoading: state.documentDetailsFront.isLoading,
});

export const DocumentCirculation = connect(
  mapStateToProps,
  null
)(DocumentCirculationComponent);
