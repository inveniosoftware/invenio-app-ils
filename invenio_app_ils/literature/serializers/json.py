# -*- coding: utf-8 -*-
#
# Copyright (C) 2018-2020 CERN.
#
# invenio-app-ils is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Literature JSON serializers."""

from invenio_records_rest.serializers.json import JSONSerializer

from .custom_fields import field_cover_metadata


class LiteratureJSONSerializer(JSONSerializer):
    """Serialize Literature."""

    def transform_record(self, pid, record, links_factory=None, **kwargs):
        """Transform record into an intermediate representation."""
        literature = super().transform_record(
            pid, record, links_factory=links_factory, **kwargs
        )
        field_cover_metadata(literature["metadata"])
        return literature

    def transform_search_hit(
        self, pid, record_hit, links_factory=None, **kwargs
    ):
        """Transform search result hit into an intermediate representation."""
        hit = super().transform_search_hit(
            pid, record_hit, links_factory=links_factory, **kwargs
        )
        field_cover_metadata(hit["metadata"])
        return hit
