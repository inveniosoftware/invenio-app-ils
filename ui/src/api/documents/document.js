import { http, apiConfig } from '../base';
import { serializer } from './serializer';
import { prepareSumQuery } from '../utils';

const documentURL = '/documents/';

const get = documentPid => {
  return http.get(`${documentURL}${documentPid}`).then(response => {
    response.data = serializer.fromJSON(response.data);
    return response;
  });
};

const create = async data => {
  const resp = await http.post(`${documentURL}`, data);
  resp.data = serializer.fromJSON(resp.data);
  return resp;
};

const update = async (docPid, data) => {
  const response = await http.put(`${documentURL}${docPid}`, data);
  response.data = serializer.fromJSON(response.data);
  return response;
};

const del = async docPid => {
  return await http.delete(`${documentURL}${docPid}`);
};

const patch = async (documentPid, ops) => {
  const response = await http.patch(`${documentURL}${documentPid}`, ops, {
    headers: { 'Content-Type': 'application/json-patch+json' },
  });
  response.data = serializer.fromJSON(response.data);
  return response;
};

const viewEvent = async docPid => {
  return await http.post(`${documentURL}${docPid}/stats`, {
    event: 'record-view',
  });
};

const createRelation = async (docPid, data) => {
  const resp = await http.post(`${documentURL}${docPid}/relations`, data);
  resp.data = serializer.fromJSON(resp.data);
  return resp;
};

const deleteRelation = async (docPid, data) => {
  const resp = await http.delete(`${documentURL}${docPid}/relations`, {
    data: data,
  });
  resp.data = serializer.fromJSON(resp.data);
  return resp;
};

class QueryBuilder {
  constructor() {
    this.overbookedQuery = [];
    this.currentlyOnLoanQuery = [];
    this.availableItemsQuery = [];
    this.withPendingLoansQuery = [];
    this.withTagQuery = [];
    this.withDocumentTypeQuery = [];
    this.withEitemsQuery = [];
    this.pendingOverdueQuery = [];
    this.withSeriesQuery = [];
    this.sortByQuery = '';
  }

  overbooked() {
    this.overbookedQuery.push('circulation.overbooked:true');
    return this;
  }

  currentlyOnLoan() {
    this.currentlyOnLoanQuery.push('circulation.active_loans:>0');
    return this;
  }

  withAvailableItems() {
    this.availableItemsQuery.push('circulation.has_items_for_loan:>0');
    return this;
  }

  withPendingLoans() {
    this.withPendingLoansQuery.push('circulation.pending_loans:>0');
    return this;
  }

  withTag(tag) {
    if (!tag) {
      throw TypeError('Tag argument missing');
    }
    this.withTagQuery.push(`tags.name:"${tag.name}"`);
    return this;
  }

  withDocumentType(documentType) {
    if (!documentType) {
      throw TypeError('documentType argument missing');
    }
    this.withDocumentTypeQuery.push(`document_type:"${documentType}"`);
    return this;
  }

  withEitems() {
    this.withEitemsQuery.push('eitems.total:>0');
    return this;
  }

  withSeriesPid(seriesPid, moi) {
    if (!seriesPid) {
      throw TypeError('Series PID argument missing');
    }
    if (moi === 'SERIAL') {
      this.withSeriesQuery.push(
        `relations.serial.pid:${prepareSumQuery(seriesPid)}`
      );
    } else {
      this.withSeriesQuery.push(
        `relations.multipart_monograph.pid:${prepareSumQuery(seriesPid)}`
      );
    }
    return this;
  }

  pendingOverdue() {
    const query = [
      'circulation.has_items_for_loan:0',
      'circulation.pending_loans:>0',
      'circulation.overdue_loans:>0',
      'items.total:>0',
    ];
    this.pendingOverdueQuery.push(encodeURI(query.join(' AND ')));
    return this;
  }

  sortBy(order = 'bestmatch') {
    this.sortByQuery = `&sort=${order}`;
    return this;
  }

  qs() {
    const searchCriteria = this.overbookedQuery
      .concat(
        this.currentlyOnLoanQuery,
        this.availableItemsQuery,
        this.withPendingLoansQuery,
        this.withTagQuery,
        this.withDocumentTypeQuery,
        this.withEitemsQuery,
        this.pendingOverdueQuery,
        this.withSeriesQuery
      )
      .join(' AND ');
    return `${searchCriteria}${this.sortByQuery}`;
  }
}

const queryBuilder = () => {
  return new QueryBuilder();
};

const list = async query => {
  const response = await http.get(`${documentURL}?q=${query}`);
  response.data.total = response.data.hits.total;
  response.data.hits = response.data.hits.hits.map(hit =>
    serializer.fromJSON(hit)
  );
  return response;
};

const count = query => {
  return http.get(`${documentURL}?q=${query}`).then(response => {
    response.data = response.data.hits.total;
    return response;
  });
};

export const document = {
  searchBaseURL: `${apiConfig.baseURL}${documentURL}`,
  get: get,
  create: create,
  update: update,
  delete: del,
  patch: patch,
  viewEvent: viewEvent,
  createRelation: createRelation,
  deleteRelation: deleteRelation,
  list: list,
  count: count,
  query: queryBuilder,
  serializer: serializer,
};
