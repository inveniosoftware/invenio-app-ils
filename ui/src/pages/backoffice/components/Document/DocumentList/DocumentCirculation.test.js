import React from 'react';
import { shallow } from 'enzyme';
import DocumentCirculation from './DocumentCirculation';
import * as testData from '@testData/documents.json';

describe('DocumentCirculation tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  const document = {
    metadata: {
      ...testData[0],
      items: { total: 2 },
      circulation: { has_items_for_loan: 2 },
    },
  };

  it('should load the DocumentCirculation component', () => {
    const component = shallow(<DocumentCirculation document={document} />);
    expect(component).toMatchSnapshot();
  });
});
