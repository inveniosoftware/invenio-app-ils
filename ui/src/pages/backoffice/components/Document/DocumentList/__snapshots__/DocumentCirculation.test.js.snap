// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentCirculation tests should load the DocumentCirculation component 1`] = `
<List
  className="document-circulation"
  verticalAlign="middle"
>
  <ListItem>
    <ListContent
      floated="right"
    >
      <strong>
        2
      </strong>
    </ListContent>
    <ListContent>
      total items
    </ListContent>
  </ListItem>
  <ListItem>
    <ListContent
      floated="right"
    >
      <strong>
        2
      </strong>
    </ListContent>
    <ListContent>
      available items
    </ListContent>
  </ListItem>
  <ListItem>
    <ListContent
      floated="right"
    >
      <strong />
    </ListContent>
    <ListContent>
      active loans
    </ListContent>
  </ListItem>
  <ListItem>
    <ListContent
      floated="right"
    >
      <strong />
    </ListContent>
    <ListContent>
      pending loans
    </ListContent>
  </ListItem>
  <ListItem>
    <ListContent
      floated="right"
    >
      <Icon
        as="i"
        color="green"
        name="minus"
        size="small"
      />
    </ListContent>
    <ListContent>
      overbooked
    </ListContent>
  </ListItem>
</List>
`;
